// plottable/build/src/core/symbolFactories.js
"use strict";
var d3 = require("d3");
function circle() {
    return function (symbolSize) { return d3.svg.symbol().type("circle").size(Math.PI * Math.pow(symbolSize / 2, 2))(null); };
}
exports.circle = circle;
function square() {
    return function (symbolSize) { return d3.svg.symbol().type("square").size(Math.pow(symbolSize, 2))(null); };
}
exports.square = square;
function cross() {
    return function (symbolSize) { return d3.svg.symbol().type("cross").size((5 / 9) * Math.pow(symbolSize, 2))(null); };
}
exports.cross = cross;
function diamond() {
    return function (symbolSize) { return d3.svg.symbol().type("diamond").size(Math.tan(Math.PI / 6) * Math.pow(symbolSize, 2) / 2)(null); };
}
exports.diamond = diamond;
function triangleUp() {
    return function (symbolSize) { return d3.svg.symbol().type("triangle-up").size(Math.sqrt(3) * Math.pow(symbolSize / 2, 2))(null); };
}
exports.triangleUp = triangleUp;
function triangleDown() {
    return function (symbolSize) { return d3.svg.symbol().type("triangle-down").size(Math.sqrt(3) * Math.pow(symbolSize / 2, 2))(null); };
}
exports.triangleDown = triangleDown;
